pygsp/graphs/nngraphs/nngraph.py

```diff
@@ -1,199 +1,560 @@
-                                 algorithm='kdtree')
-
-            else:
-                kdt = spatial.KDTree(Xout)
-                D, NN = kdt.query(Xout, k=(k + 1),
-                                  p=dist_translation[dist_type])
-
-            if self.sigma is None:
-                self.sigma = np.mean(D[:, 1:])  # Discard distance to self.
-
-            for i in range(N):
-                spi[i * k:(i + 1) * k] = np.kron(np.ones((k)), i)
-                spj[i * k:(i + 1) * k] = NN[i, 1:]
-                spv[i * k:(i + 1) * k] = np.exp(-np.power(D[i, 1:], 2) /
-                                                float(self.sigma))
-
-        elif self.NNtype == 'radius':
-
-            kdt = spatial.KDTree(Xout)
-            D, NN = kdt.query(Xout, k=None, distance_upper_bound=epsilon,
-                              p=dist_translation[dist_type])
-            if self.sigma is None:
-                # Discard distance to self.
-                self.sigma = np.mean([np.mean(d[1:]) for d in D])
-            count = 0
-            for i in range(N):
-                count = count + len(NN[i])
-
-            spi = np.zeros((count))
-            spj = np.zeros((count))
-            spv = np.zeros((count))
-
-            start = 0
-            for i in range(N):
-                leng = len(NN[i]) - 1
-                spi[start:start + leng] = np.kron(np.ones((leng)), i)
-                spj[start:start + leng] = NN[i][1:]
-                spv[start:start + leng] = np.exp(-np.power(D[i][1:], 2) /
-                                                 float(self.sigma))
-                start = start + leng
+    Choice of kernel.
 
-        else:
-            raise ValueError('Unknown NNtype {}'.format(self.NNtype))
+    >>> fig, axes = plt.subplots(1, 2)
+    >>> width = 0.3
+    >>> distances = np.linspace(0, 1, 200)
+    >>> for name, kernel in graphs.NNGraph._kernels.items():
+    ...     _ = axes[0].plot(distances, kernel(distances / width), label=name)
+    >>> _ = axes[0].set_xlabel('distance [0, inf]')
+    >>> _ = axes[0].set_ylabel('similarity [0, 1]')
+    >>> _ = axes[0].legend(loc='upper right')
+    >>> features = rs.uniform(size=(100, 3))
+    >>> for kernel in ['gaussian', 'triangular', 'tricube', 'exponential']:
+    ...     G = graphs.NNGraph(features, kernel=kernel)
+    ...     _ = axes[1].hist(G.W.data, bins=20, label=kernel, alpha=0.5)
+    >>> _ = axes[1].legend()
+    >>> _ = axes[1].set_title('edge weights')
+
+    Choice of kernel width.
+
+    >>> fig, axes = plt.subplots()
+    >>> for width in [.2, .3, .4, .6, .8, None]:
+    ...     G = graphs.NNGraph(features, kernel_width=width)
+    ...     label = 'width = {:.2f}'.format(G.kernel_width)
+    ...     _ = axes.hist(G.W.data, bins=20, label=label, alpha=0.5)
+    >>> _ = axes.legend(loc='upper left')
+    >>> _ = axes.set_title('edge weights')
+
+    Choice of backend. Compare on your data!
+
+    >>> import time
```

Comment:
Is this executed by travis? If so, should we make sure it does not take too much time?

Reply:
It is. Execution time is fine from my testing.